pkg/controller/federatedcluster/controller.go

```diff
@@ -160,82 +150,146 @@ func (cc *ClusterController) Run(stopChan <-chan struct{}) {
 		if err := cc.updateClusterStatus(); err != nil {
 			glog.Errorf("Error monitoring cluster status: %v", err)
 		}
-	}, cc.clusterMonitorPeriod, stopChan)
+	}, cc.clusterHealthCheckConfig.PeriodSeconds, stopChan)
 }
 
-// updateClusterStatus checks cluster status and get the metrics from cluster's restapi
+// updateClusterStatus checks cluster health and updates status of all FederatedClusters
 func (cc *ClusterController) updateClusterStatus() error {
 	clusters := &fedv1a1.FederatedClusterList{}
 	err := cc.client.List(context.TODO(), clusters, cc.fedNamespace)
 	if err != nil {
 		return err
 	}
 
-	for _, cluster := range clusters.Items {
+	var wg sync.WaitGroup
+	for _, obj := range clusters.Items {
 		cc.mu.RLock()
-		// skip updating status of the cluster which is not yet added to knownClusterSet.
-		if !cc.knownClusterSet.Has(cluster.Name) {
-			cc.mu.RUnlock()
+		cluster := obj.DeepCopy()
+		clusterStore := cc.clusterStoreMap[cluster.Name]
+		cc.mu.RUnlock()
+		if clusterStore == nil {
+			glog.Warningf("Failed to retrieve stored data for cluster %s", cluster.Name)
 			continue
 		}
-		clusterClient, clientFound := cc.clusterKubeClientMap[cluster.Name]
-		clusterStatusOld, statusFound := cc.clusterStatusMap[cluster.Name]
-		cc.mu.RUnlock()
 
-		if !clientFound {
-			glog.Warningf("Failed to get client for cluster %s", cluster.Name)
-			continue
+		wg.Add(1)
+		go cc.updateIndividualClusterStatus(cluster, clusterStore, &wg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (cc *ClusterController) updateIndividualClusterStatus(cluster *fedv1a1.FederatedCluster,
+	store *ClusterStore, wg *sync.WaitGroup) {
+	clusterClient := store.clusterKubeClient
+
+	currentClusterStatus := thresholdAdjustedClusterStatus(clusterClient.GetClusterHealthStatus(), store,
```

Comment:
I think `clusterClient.GetClusterHealthStatus()` should be on its own line.  It is not a trivial call, and having it on its own line simplifies debugging and avoids obscuring the importance of the call.

Reply:
Done